### PR TITLE
Legacy device facade class

### PR DESCRIPTION
**Pull request by albi3ro**
**Context:**

We currently have to maintain logic for interacting with two separate device interfaces.  This adds extra maintainence, and creates a lot more paths where things can go wrong or get confusing.

Given the number of low-priority plugins that follow the legacy device interface that we probably won't get around to updating, we probably won't be able to get rid of the legacy device interface any time soon.  But we can isolate it and limit it's scope.

**Description of the Change:**

Adds a `LegacyDeviceFacade` class that maps a device from the old device interface to the new device interface.

**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:**

[sc-58353]

### CODE REVIEW

**Review by Shiro-Raven — APPROVED** [review]
LGTM! Next up is figuring out how to keep testing for device switching when DQ.Legacy is gone

File: pennylane/devices/legacy_facade.py [hunk 1]
```diff
@@ -0,0 +1,428 @@
...
+        # create new backprop device
+        expand_fn = self._device.expand_fn
+        batch_transform = self._device.batch_transform
+        if hasattr(self._device, "_debugger"):
+            debugger = self._device._debugger
+        else:
+            debugger = "No debugger"
+        tracker = self._device.tracker
+
+        backprop_devices = self._device.capabilities().get("passthru_devices", None)
```
Comment (mudit2812) [inline pennylane/devices/legacy_facade.py:315]: You already initialized `backprop_devices` above.

```suggestion
        tracker = self._device.tracker
```

File: pennylane/devices/legacy_facade.py [hunk 2]
```diff
@@ -0,0 +1,428 @@
...
+        new_device.tracker = tracker
+
+        return new_device
+
+    # pylint: disable=protected-access
+    def _update_original_device(self, temp_device):
+        """After performing an execution with a backprop device, update the state of the original device."""
+        # Update for state vector simulators that have the _pre_rotated_state attribute
+        if hasattr(self._device, "_pre_rotated_state"):
+            self._device._pre_rotated_state = temp_device._pre_rotated_state
```
Comment (mudit2812) [inline pennylane/devices/legacy_facade.py:333]: Won't this introduce interface data into the original device?

File: pennylane/devices/legacy_facade.py [hunk 3]
```diff
@@ -0,0 +1,428 @@
...
+
+        # Update for state vector simulators that have the _state attribute
+        if hasattr(self._device, "_state"):
+            self._device._state = temp_device._state
+
+    def _validate_backprop_method(self, tape):
+        print("in _valdiate_backprop_method")
+        if tape.shots:
+            print("has shots")
```
Comment (mudit2812) [inline pennylane/devices/legacy_facade.py:342]:
```suggestion
```

File: pennylane/devices/legacy_facade.py [hunk 4]
```diff
@@ -0,0 +1,428 @@
...
+            interface = None
+        mapped_interface = qml.workflow.execution.INTERFACE_MAP.get(interface, interface)
+
+        # determine if the device supports backpropagation
+        backprop_interface = self._device.capabilities().get("passthru_interface", None)
+        print("im here")
+        if backprop_interface is not None:
+            # device supports backpropagation natively
+            print("backprop_interface", backprop_interface)
+            print("mapped_interface: ", mapped_interface)
```
Comment (mudit2812) [inline pennylane/devices/legacy_facade.py:360]:
```suggestion
        if tape.shots:
            return False
        params = tape.get_parameters(trainable_only=False)
        interface = qml.math.get_interface(*params)

        if tape and any(isinstance(m.obs, qml.SparseHamiltonian) for m in tape.measurements):
            return False
        if interface == "numpy":
            interface = None
        mapped_interface = qml.workflow.execution.INTERFACE_MAP.get(interface, interface)

        # determine if the device supports backpropagation
        backprop_interface = self._device.capabilities().get("passthru_interface", None)
        if backprop_interface is not None:
            # device supports backpropagation natively
```

File: tests/devices/test_legacy_facade.py [hunk 5]
```diff
@@ -0,0 +1,447 @@
...
+from pennylane.devices.execution_config import ExecutionConfig
+from pennylane.devices.legacy_facade import (
+    LegacyDeviceFacade,
+    legacy_device_batch_transform,
+    legacy_device_expand_fn,
+)
+
+
+class DummyDevice(qml.devices.LegacyDevice):
+    """A minimal device that does not do"""
```
Comment (mudit2812) [inline tests/devices/test_legacy_facade.py:33]: Could you update this docstring?

File: pennylane/devices/legacy_facade.py (hunk 2 above)
Comment (albi3ro) [reply pennylane/devices/legacy_facade.py:333]: Just copying code from the qnode.  Assuming it was in the qnode for some reason.

File: pennylane/devices/legacy_facade.py (hunk 4 above)
Comment (albi3ro) [reply pennylane/devices/legacy_facade.py:360]: 😅

File: tests/devices/test_legacy_facade.py (hunk 5 above)
Comment (albi3ro) [reply tests/devices/test_legacy_facade.py:33]:
```suggestion
    """A minimal device that does not do anything."""
```

**Review by mudit2812 — APPROVED** [review]

**Pull request merged**